src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs

```diff
@@ -37,6 +38,36 @@ public override int GetHashCode()
         {
             return Hash.Combine(Input.GetHashCode(), this.Symbol?.GetHashCode() ?? 0);
         }
+
+#if DEBUG
+        private int _id = -1;
+        private bool _idWasRead;
+
+        public int Id
+        {
+            get
+            {
+                if (_id != -1)
+                {
+                    _idWasRead = true;
+                }
+                return _id;
+            }
+            internal set
+            {
+                Debug.Assert(!_idWasRead, "Id was set after reading it");
```

Comment:
Since "set after read" checks only occur after a set to a non-negative Id, they're probably redundant with the "double set" check.

Reply:
I agree. We could remove `_isWasRead` now that uninitialized value is `-1`.

Reply:
Done